service/resource/pod/delete.go

```diff
@@ -2,15 +2,116 @@ package pod
 
 import (
 	"context"
+	"time"
 
+	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/operatorkit/framework"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apismetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+
+	"github.com/giantswarm/kvm-operator/service/key"
+)
+
+const (
+	PodDeletionGracePeriod = 0 * time.Second
 )
 
 func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error {
+	podToDelete, err := key.ToPod(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if podToDelete != nil {
+		{
+			r.logger.LogCtx(ctx, "debug", "updating the pod in the Kubernetes API to remove the pod's 'draining-nodes' finalizer")
+
+			_, err := r.k8sClient.CoreV1().Pods(podToDelete.Namespace).Update(podToDelete)
+			if apierrors.IsConflict(err) {
+				// The reconciled pod may be updated by other processes or even humans
+				// meanwhile. In case the resource version we currently know does not
+				// match the latest existing one, we give up here and wait for the
+				// delete event to be replayed. Then we try again later until we
+				// succeed.
+				r.logger.LogCtx(ctx, "debug", "cannot update the pod in the Kubernetes API to remove the pod's 'draining-nodes' finalizer because of outdated resource version")
+				return nil
+			} else if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "debug", "updated the pod in the Kubernetes API to remove the pod's 'draining-nodes' finalizer")
+		}
+
+		{
+			r.logger.LogCtx(ctx, "debug", "deleting the pod in the Kubernetes API")
+
+			gracePeriodSeconds := int64(PodDeletionGracePeriod.Seconds())
+			options := &apismetav1.DeleteOptions{
+				GracePeriodSeconds: &gracePeriodSeconds,
+			}
+			err = r.k8sClient.CoreV1().Pods(podToDelete.Namespace).Delete(podToDelete.Name, options)
+			if err != nil {
+				return microerror.Mask(err)
+			}
```

Comment:
For my understanding: do we still have to call delete once again after removing finalizer? 
Is that you have to make delete request at least `len(finalizers)+1` times in order to remove object?

I thought it looks like:
1. Pod is deleted.
2. Delete evnet arrives.
3. kvm-operator removes the finalizer.
4. Pod is deleted automatically as there are no finalizers left.

Reply:
There are two different things we want to control. The container termination and the pod object garbage collection. We create the pods now with a grace period of 5 minutes. This delays the container termination to be able to drain the guest cluster nodes. The finalizer is only to replay the delete event. Even if the finalizer is removed, the grace period is still active. We delete with a grace period of zero here to overwrite the pod definition like doing `kubectl delete --force --grace-period=0`.

Reply:
Ok, so the pod would be removed after grace period if the second delete wasn't requested, correct?

Why we need to create pods with grace period when we use finalizers? Doesn't the operator do the draining and decides when to remove the finalizer and remove the pod? Why is it 5 minutes?

Reply:
We want to be able to act on delete events. On delete events the grace period is 30 seconds by default. This might not be enough for us to be able to properly drain guest cluster nodes. That is why we extend it to something big like 5 minutes. After some time we have to reap everything anyway. Again, finalizers and grace period do different things. Finalizers are only to replay delete events so we can ensure the deletion is managed properly until we reach the desired delete state. The grace period is just to not kill everything immediately so we can drain.

Reply:
Ok, so after calling `kubectl delete pod A`, `kubectl get pod A` shows nothing even if `A` had a finalizer on it?
Then the delete event is replayed as long as something removes the finalizer?

Reply:
In case it had have a finalizer the pod object would not be deleted but the KILL signal would have sent to the containers within the pod.

Reply:
💡 Thanks!